substrate/frame/support/src/traits/storage.rs

```diff
@@ -127,6 +132,69 @@ impl WhitelistedStorageKeys for Tuple {
 	}
 }
 
+/// The resource footprint of a bunch of blobs. We assume only the number of blobs and their total
+/// size in bytes matter.
+#[derive(Default, Copy, Clone, Eq, PartialEq, RuntimeDebug)]
+pub struct Footprint {
+	/// The number of blobs.
+	pub count: u64,
+	/// The total size of the blobs in bytes.
+	pub size: u64,
+}
+
+impl Footprint {
+	pub fn from_parts(items: usize, len: usize) -> Self {
+		Self { count: items as u64, size: len as u64 }
+	}
+
+	pub fn from_encodable(e: impl Encode) -> Self {
+		Self::from_parts(1, e.encoded_size())
+	}
+}
+
+/// Some sort of cost taken from account temporarily in order to offset the cost to the chain of
+/// holding some data `Footprint` in state.
+///
+/// The cost may be increased, reduced or dropped entirely as the footprint changes.
+pub trait Consideration<AccountId> {
+	/// A single ticket corresponding to some particular datum held in storage. This is an opaque
+	/// type, but must itself be stored and generally it should be placed alongside whatever data
+	/// the ticket was created for.
+	///
+	/// While not technically a linear type owing to the need for `FullCodec`, *this should be
+	/// treated as one*. Don't type to duplicate it, and remember to drop it when you're done with
+	/// it.
+	type Ticket: Member + FullCodec + TypeInfo + MaxEncodedLen + Default;
+
+	/// Optionally consume an old ticket and alter the footprint, enforcing the new cost to `who`
+	/// and returning the new ticket (or an error if there was an issue).
+	///
+	/// For creating tickets and dropping them, you can use the simpler `new` and `drop` instead.
+	fn update(
+		who: &AccountId,
+		old: Option<Self::Ticket>,
+		new: Option<Footprint>,
+	) -> Result<Self::Ticket, DispatchError>;
+
+	/// Create a ticket for the `new` footprint attributable to `who`. This ticket *must* be
+	/// consumed (through either `drop` or `update`) if the footprint changes or is removed.
+	fn new(who: &AccountId, new: Footprint) -> Result<Self::Ticket, DispatchError> {
```

Comment:
Looks like this function would benefit from the [`#[must_use]` attribute](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute)?